libgnucash/engine/Account.cpp

```diff
@@ -5417,120 +5419,155 @@ gnc_account_imap_delete_account (GncImportMatchMap *imap,
+ *
+ * This approach is enhanced by a token selection algorithm. This algorithm
+ * determines, which tokens should be used for calculating the averaged account
+ * probabilities and which tokens should be excluded.
  */
-struct AccountProbability
-{
-    double product; /* product of probabilities */
-    double product_difference; /* product of (1-probabilities) */
-};
 
-struct AccountTokenCount
+struct AccountTokenProbability
 {
     std::string account_guid;
     int64_t token_count; /** occurrences of a given token for this account_guid */
+    double probability; /* conditional probability of this account_guid for a given token */
 };
 
 /** total_count and the token_count for a given account let us calculate the
  * probability of a given account with any single token
  */
 struct TokenAccountsInfo
 {
-    std::vector<AccountTokenCount> accounts;
+    std::vector<AccountTokenProbability> accounts;
     int64_t total_count;
 };
 
-/** holds an account guid and its corresponding integer probability
-  the integer probability is some factor of 10
- */
-struct AccountInfo
-{
-    std::string account_guid;
-    int32_t probability;
-};
-
 static void
 build_token_info(char const * suffix, KvpValue * value, TokenAccountsInfo & tokenInfo)
 {
     if (strlen(suffix) == GUID_ENCODING_LENGTH)
     {
         tokenInfo.total_count += value->get<int64_t>();
         /*By convention, the key ends with the account GUID.*/
-        tokenInfo.accounts.emplace_back(AccountTokenCount{std::string{suffix}, value->get<int64_t>()});
+        tokenInfo.accounts.emplace_back(AccountTokenProbability{std::string{suffix}, value->get<int64_t>(), 0.0});
     }
 }
 
-/** We scale the probability values by probability_factor.
-  ie. with probability_factor of 100000, 10% would be
-  0.10 * 100000 = 10000 */
-static constexpr int probability_factor = 100000;
-
-static FinalProbabilityVec
-build_probabilities(ProbabilityVec const & first_pass)
+static TokenInfoVec
+get_token_info(GncImportMatchMap * imap, GList * tokens)
 {
-    FinalProbabilityVec ret;
-    for (auto const & first_pass_prob : first_pass)
+    TokenInfoVec ret;
+    for (auto current_token = tokens; current_token; current_token = current_token->next)
     {
-        auto const & account_probability = first_pass_prob.second;
-        /* P(AB) = A*B / [A*B + (1-A)*(1-B)]
-         * NOTE: so we only keep track of a running product(A*B*C...)
-         * and product difference ((1-A)(1-B)...)
-         */
-        int32_t probability = (account_probability.product /
-                (account_probability.product + account_probability.product_difference)) * probability_factor;
-        ret.push_back({first_pass_prob.first, probability});
+        /* query token frequencies from import match map */
+        TokenAccountsInfo token_info{};
+        std::string token_name{static_cast <char const*> (current_token->data)};
+        auto path = std::string(IMAP_FRAME_BAYES "/") + token_name + "/";
```

Comment:
Reasoning about this code a bit more, `token_name` is a string created in line 5490 and then 5491, there is a new string created, and the only time, it's used is in 5499 which is done only if `token_info.total_count > 0`.

Can the string object creation be postponed till needed?

How about?
```
auto token_name = static_cast<char const *>(current_token->data);
auto path = std::string(IMAP_FRAME_BAYES "/").append(token_name).append("/");```
```

Line 5499 should be left as it is, since string is implicitly convertible from `const char *`

Reply:
What's the difference between [`std::operator+(std::string)`](https://en.cppreference.com/w/cpp/string/basic_string/operator%2B) and [`std::string::append()`](https://en.cppreference.com/w/cpp/string/basic_string/append)?

Reply:
String object creation postponed as suggested